src/Access/SettingsConstraints.cpp

```diff
@@ -269,6 +297,58 @@ bool SettingsConstraints::checkImpl(const Settings & current_settings, SettingCh
     }
 
     const Constraint * constraint = tryGetConstraint(setting_name);
+    if (current_settings.readonly == 1)
+    {
+        if (!constraint || (constraint->min_value_in_readonly.isNull() && constraint->max_value_in_readonly.isNull())) {
+            if (reaction == THROW_ON_VIOLATION)
+                throw Exception("Cannot modify '" + setting_name + "' setting in readonly mode", ErrorCodes::READONLY);
+            else
+                return false;
+        }
+
+        const Field & min_value = constraint->min_value_in_readonly;
+        const Field & max_value = constraint->max_value_in_readonly;
+        if (!min_value.isNull() && !max_value.isNull() && less_or_cannot_compare(max_value, min_value))
+        {
+            if (reaction == THROW_ON_VIOLATION)
+                throw Exception("Cannot modify '" + setting_name + "' setting in readonly mode", ErrorCodes::READONLY);
+            else
+                return false;
+        }
+
+        // Check left bound
+        if (!min_value.isNull() && less_or_cannot_compare(new_value, min_value))
+        {
+            if (reaction == THROW_ON_VIOLATION)
+            {
+                throw Exception(
+                    "Setting " + setting_name + " in readonly mode shouldn't be less than " + applyVisitor(FieldVisitorToString(), constraint->min_value_in_readonly),
+                    ErrorCodes::READONLY);
```

Comment:
Just style, but I'd prefer use format string in exceptions
```suggestion
                throw Exception(ErrorCodes::READONLY,
                    "Setting '{}' in readonly mode shouldn't be less than {}",
                    setting_name, applyVisitor(FieldVisitorToString(), constraint->min_value_in_readonly));
```

Reply:
I agree, but it is copy-pasted and kept to be consistent throughout the file